pkg/pillar/cmd/nim/controllerdns.go

```diff
@@ -70,28 +71,78 @@ func (n *nim) queryControllerDNS() {
 	}
 }
 
-// periodical cache the controller DNS resolution into /etc/hosts file
-// it returns the cached ip string, and TTL setting from the server
-func (n *nim) controllerDNSCache(etchosts, controllerServer []byte, ipaddrCached string) (string, int) {
-	if len(etchosts) == 0 || len(controllerServer) == 0 {
-		return ipaddrCached, maxTTLSec
+func (n *nim) resolveWithPorts(domain string) []devicenetwork.DNSResponse {
+	dnsResponse, errs := devicenetwork.ResolveWithPortsLambda(
+		domain,
+		n.dpcManager.GetDNS(),
+		devicenetwork.ResolveWithSrcIP,
+	)
+	if len(errs) > 0 {
+		n.Log.Warnf("resolveWithPortsLambda failed: %+v", errs)
 	}
+	return dnsResponse
+}
 
+// periodical cache the controller DNS resolution into /etc/hosts file
+// it returns the cached ip string, and TTL setting from the server
+func (n *nim) controllerDNSCache(
+	etchosts, controllerServer []byte,
+	ipaddrCached string,
+) (string, int) {
 	// Check to see if the server domain is already in the /etc/hosts as in eden,
 	// then skip this DNS queries
-	if ipaddrCached == "" {
-		hostsEntries := bytes.Split(etchosts, []byte("\n"))
-		for _, entry := range hostsEntries {
-			fields := bytes.Fields(entry)
-			if len(fields) == 2 {
-				if bytes.Compare(fields[1], controllerServer) == 0 {
-					n.Log.Tracef("server entry %s already in /etc/hosts, skip", controllerServer)
-					return ipaddrCached, maxTTLSec
-				}
-			}
-		}
+	isCached, ipAddrCached, ttlCached := n.checkCachedEntry(
+		etchosts,
+		controllerServer,
+		ipaddrCached,
+	)
+	if isCached {
+		return ipAddrCached, ttlCached
+	}
+
+	err := os.Remove(tmpHostFileName)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		n.Log.Warnf("%s exists but removing failed: %+v", tmpHostFileName, err)
+	}
+
+	var newhosts []byte
+	var ttlSec int
+	var lookupIPaddr string
+
+	dnsResponses := n.resolveWithPorts(string(controllerServer))
+	ttlSec = int(dnsResponses[0].TTL)
+	lookupIPaddr = dnsResponses[0].IP.String()
```

Comment:
Finally see the complete panic stacktrace from the ztests run and it says
2023-03-30T17:12:55Z,pillar;panic: runtime error: index out of range [0] with length 0
2023-03-30T17:12:55Z,pillar;
2023-03-30T17:12:55Z,pillar;goroutine 512 [running]:
2023-03-30T17:12:55Z,pillar;github.com/lf-edge/eve/pkg/pillar/cmd/nim.(*nim).controllerDNSCache(0xc00035f400, {0xc000c1b200, 0xa0, 0x200}, {0xc000fbf200, 0x19, 0x200}, {0xc002376d60?, 0xc1018bbeb05b9c8e?})
2023-03-30T17:12:55Z,pillar;	/pillar/cmd/nim/controllerdns.go:113 +0x387
2023-03-30T17:12:55Z,pillar;github.com/lf-edge/eve/pkg/pillar/cmd/nim.(*nim).queryControllerDNS(0xc00035f400)
2023-03-30T17:12:55Z,pillar;	/pillar/cmd/nim/controllerdns.go:65 +0x3d6
2023-03-30T17:12:55Z,pillar;created by github.com/lf-edge/eve/pkg/pillar/cmd/nim.(*nim).run.func1
2023-03-30T17:12:55Z,pillar;	/pillar/cmd/nim/nim.go:316 +0x196

Instead of the various checks (and these missing ones) for len(dnsResponses) == 0, it would be easier to read if there was a single if (len(dnsResponses) == 0 block here.

Also, if we get multiple responses shouldn't we use all of them? (I didn't check that the old dns code did when multiple responses for a name).